codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java

```diff
@@ -66,412 +72,1106 @@
+        }
+
+        void streamClosed() {
+            streamClosedWithoutError = true;
+            if (readInProgress) {
+                // Just call closeForcibly() as this will take care of fireChannelInactive().
+                unsafe().closeForcibly();
+            } else {
+                closePending = true;
             }
-            super.doClose();
+            ((Http2MultiplexCodecStream) stream).channel = null;
         }
 
         @Override
-        protected void doWrite(Object msg) {
+        public ChannelMetadata metadata() {
+            return METADATA;
+        }
 
-            if (msg instanceof Http2StreamFrame) {
-                Http2StreamFrame frame = (Http2StreamFrame) msg;
-                ChannelPromise promise = ctx.newPromise();
-                if (isStreamIdValid(frame.streamId())) {
-                    ReferenceCountUtil.release(frame);
-                    throw new IllegalArgumentException("Stream id must not be set on the frame. Was: "
-                        + frame.streamId());
-                }
-                if (!isStreamIdValid(streamId())) {
-                    if (!(frame instanceof Http2HeadersFrame)) {
-                        ReferenceCountUtil.release(frame);
-                        throw new IllegalArgumentException("The first frame must be a headers frame. Was: "
-                            + frame.name());
-                    }
-                    frame = new ChannelCarryingHeadersFrame((Http2HeadersFrame) frame, this);
-                    // Handle errors on stream creation
-                    promise.addListener(this);
-                } else {
-                    frame.streamId(streamId());
-                }
-                writeFromStreamChannel(frame, promise, false);
-            } else if (msg instanceof Http2GoAwayFrame) {
-                ChannelPromise promise = ctx.newPromise();
-                promise.addListener(this);
-                writeFromStreamChannel(msg, promise, false);
-            } else {
-                ReferenceCountUtil.release(msg);
-                throw new IllegalArgumentException("Message must be an Http2GoAwayFrame or Http2StreamFrame: " + msg);
-            }
+        @Override
+        public ChannelConfig config() {
+            return config;
         }
 
         @Override
-        protected void doWriteComplete() {
-            flushFromStreamChannel();
+        public boolean isOpen() {
+            return !closePromise.isDone();
         }
 
         @Override
-        protected EventExecutor preferredEventExecutor() {
-            return ctx.executor();
+        public boolean isActive() {
+            return isOpen();
         }
 
         @Override
-        protected void bytesConsumed(final int bytes) {
-            ctx.write(new DefaultHttp2WindowUpdateFrame(bytes).streamId(streamId()));
+        public boolean isWritable() {
+            // TODO: Should we also take the parent channel into account ?
```

Comment:
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1135)